File: ruby/lib/synt/similar/jaccard.rb
#  See: http://en.wikipedia.org/wiki/Jaccard_index
require 'manowar'

define 'Synt::Similar'

module Synt::Similar::Jaccard
  extend self

  def compare src, cmp
    a = src.uniq
    b = cmp.uniq
    i = a & b
    u = a | b
    i.length.to_f / u.length.to_f * 100
  end
end

File: haskell/src/Synt/Similar/Jaccard.hs
module Synt.Similar.Jaccard (jaccard) where

import Data.List

float :: Int -> Float
float = fromIntegral

jaccard :: [String] -> [String] -> Float
jaccard a b = do
  let i = a `intersect` b
  let u = a `union` b
  float (length i) / float (length u) * 100
